src/snowflake/snowpark/_internal/debug_utils.py

```diff
@@ -0,0 +1,109 @@
+#
+# Copyright (c) 2012-2025 Snowflake Computing Inc. All rights reserved.
+#
+
+from functools import cached_property
+import os
+
+from snowflake.snowpark._internal.ast.batch import get_dependent_bind_ids
+from snowflake.snowpark._internal.ast.utils import __STRING_INTERNING_MAP__
+
+UNKNOWN_FILE = "__UNKNOWN_FILE__"
+
+
+class DataFrameLineageNode:
+    """A node representing a dataframe operation in the DAG that represents the lineage of a DataFrame."""
+
+    def __init__(self, batch_id: int, stmt_cache) -> None:
+        self.batch_id = batch_id
+        self.stmt_cache = stmt_cache
+
+    @cached_property
+    def children(self) -> set[int]:
+        """Returns the batch_ids of the children of this node."""
+        return get_dependent_bind_ids(self.stmt_cache[self.batch_id])
+
+    def get_src(self):
+        """The source Stmt of the DataFrame described by the batch_id."""
+        stmt = self.stmt_cache[self.batch_id]
+        api_call = stmt.bind.expr.WhichOneof("variant")
+        return (
+            getattr(stmt.bind.expr, api_call).src
+            if api_call and getattr(stmt.bind.expr, api_call).HasField("src")
+            else None
+        )
+
+    def _read_file(
+        self, filename, start_line, end_line, start_column, end_column
+    ) -> str:
+        """Read the relevant code snippets of where the DataFrame was created. The filename given here
+        must have read permissions for the executing user."""
+        with open(filename) as f:
+            code_lines = []
+            for i, line in enumerate(f, 1):
+                if end_line == 0:
+                    if i == start_line:
+                        code_lines.append(line)
```

Comment:
I just want to clarify that if end_line is 0 (I'm assuming 0 means first line), then start_line and i must also be 0 right?

Reply:
I should improve the comments here. end_line = 0 in python 3.9/3.10 environments. In those cases, we only know that start line. Therefore we have a distinction based on end_line. I think I should redo the if statement using 
```
if sys.version >= (3, 11):
    # condition 1
else:
    # condition 2
```